specs/rollup-node-p2p.md

```diff
@@ -0,0 +1,258 @@
+  i.e. `SHA256(MESSAGE_DOMAIN_VALID_SNAPPY + snappy_decompress(message.data))[:20]`.
+- Otherwise, set `message-id` to the first 20 bytes of the `SHA256` hash of
+  the concatenation of `MESSAGE_DOMAIN_INVALID_SNAPPY` with the raw message data,
+  i.e. `SHA256(MESSAGE_DOMAIN_INVALID_SNAPPY + message.data)[:20]`.
+
+#### Heartbeat and parameters
+
+GossipSub [parameters][gossip-parameters]:
+
+- `D` (topic stable mesh target count): 8
+- `D_low` (topic stable mesh low watermark): 6
+- `D_high` (topic stable mesh high watermark): 12
+- `D_lazy` (gossip target): 6
+- `heartbeat_interval` (frequency of heartbeat, seconds): 0.5
+- `fanout_ttl` (ttl for fanout maps for topics we are not subscribed to but have published to, seconds): 24
+- `mcache_len` (number of windows to retain full messages in cache for `IWANT` responses): 12
+- `mcache_gossip` (number of windows to gossip about): 3
+- `seen_ttl` (number of heartbeat intervals to retain message IDs): 40
+
+Notable differences from L1 consensus (Eth2):
+
+- `seen_ttl` does not need to cover a full L1 epoch (6.4 minutes), but rather just a small window covering latest blocks
+- `fanout_ttl`: adjusted to lower than `seen_ttl`
+- `mcache_len`: a larger number of heartbeats can be retained since the gossip is much less noisy.
+- `heartbeat_interval`: faster interval to reduce latency, bandwidth should still be reasonable since
+  there are far fewer messages to gossip about each interval than on L1 which uses an interval of 0.7 seconds.
+
+#### Topic configuration
+
+Topics have string identifiers and are communicated with messages and subscriptions.
+`/optimism/chain_id/hardfork_version/Name`
+
+- `chain_id`: replace with decimal representation of chain ID
+- `hardfork_version`: replace with decimal representation of hardfork
+- `Name`: topic application-name
+
+Note that the topic encoding depends on the topic, unlike L1,
+since there are less topics, and all are snappy-compressed.
+
+## Gossip Topics
+
+### `blocks`
+
+The primary topic of the L2, to distribute blocks to other nodes faster than proxying through L1 would.
+
+#### Block encoding
+
+TODO: encode execution payload (SSZ or RLP), with sequencer identifier and signature.
+
+TODO: signature type and verification (`secp256k1` like transactions, with some different signature domain)
+
+#### Block validation
+
+To ensure malicious peers get scored based on application behavior the validation signals
+`ACCEPT` (valid), `IGNORE` (like inactivity) or `REJECT` (score penalties).
+
+In order of operation:
+
+- `[REJECT]` if the encoding or compression is not valid
+- `[REJECT]` if the block timestamp is older than 20 seconds in the past
+  (graceful boundary for worst-case propagation and clock skew)
+- `[REJECT]` if the block timestamp is more than 5 seconds into the future (graceful boundary for clock skew)
+- `[REJECT]` if the signature by the sequencer is not valid
```

Comment:
How do we manage the set of sequencer keys? Does this live in the L1 state?

Reply:
Undecided yet, we could put a pubkey (or even just an address, since the pubkey can be recovered from the sig) in the L1 state and read it with an RPC call yes. Or we just make it part of the chain configuration while we don't have more than 1 sequencer yet. Maybe needs a key-replace mechanism though.